infoschema/tables.go

```diff
@@ -2003,6 +2004,37 @@ func GetDataFromSessionVariables(ctx context.Context, sctx sessionctx.Context) (
 	return rows, nil
 }
 
+// GetDataFromSessionConnectAttrs produces the rows for the session_connect_attrs table.
+func GetDataFromSessionConnectAttrs(sctx sessionctx.Context) ([][]types.Datum, error) {
+	sm := sctx.GetSessionManager()
+	if sm == nil {
+		return nil, nil
+	}
+	allAttrs := sm.GetConAttrs()
+	rows := make([][]types.Datum, 0, len(allAttrs)*10) // 10 Attributes per connection
```

Comment:
Why here set 10 Attributes? Can we use actual Attributes number column?

Reply:
As `allAttrs` is a map of maps we would need to loop over all maps and get the sum of the lengths of all maps. Every map is one connection and often there are about 6 attributes per connection. So getting the exact map takes more CPU time while the less exact number might allocate more rows than needed, or would under allocate if there are many rows per connection.

```go
package main

import "testing"

func setup() map[uint64]map[string]string {
	m := make(map[uint64]map[string]string)
	for i := uint64(0); i < 100000; i++ {
		m[i] = map[string]string{
			"foo1": "bar",
			"foo2": "bar",
			"foo3": "bar",
			"foo4": "bar",
			"foo5": "bar",
			"foo6": "bar",
		}
	}
	return m
}

func BenchmarkSimple(b *testing.B) {
	m := setup()
	b.ResetTimer()
	println("Simple length of m: ", len(m)*10)
}

func BenchmarkExact(b *testing.B) {
	m := setup()
	b.ResetTimer()
	var t int
	for _, k := range m {
		t = t + len(k)
	}
	println("Exact length of m: ", t)
}
```

```
$ go test -bench .
Simple length of m:  1000000
goos: linux
goarch: amd64
pkg: foo
cpu: Intel(R) Core(TM) i7-10510U CPU @ 1.80GHz
BenchmarkSimple-8   	Simple length of m:  1000000
Simple length of m:  1000000
Simple length of m:  1000000
Simple length of m:  1000000
Simple length of m:  1000000
1000000000	         0.0000109 ns/op
Exact length of m:  600000
BenchmarkExact-8    	Exact length of m:  600000
Exact length of m:  600000
Exact length of m:  600000
Exact length of m:  600000
Exact length of m:  600000
1000000000	         0.003883 ns/op
PASS
ok  	foo	0.625s
```

Reply:
I think both methods are fine. Should I change the code to use the more exact one?